jacodb-panda-dynamic/src/main/kotlin/org/jacodb/panda/dynamic/ets/dto/Values.kt

```diff
@@ -0,0 +1,421 @@
+// data class ArrayLiteral(
+//     val elements: List<Value>,
+//     override val type: ArrayType,
+// ) : Constant
+//
+// @Serializable
+// @SerialName("ObjectLiteral")
+// data class ObjectLiteral(
+//     val keys: List<String>,
+//     val values: List<Value>,
+//     override val type: Type,
+// ) : Constant
+
+@Serializable
+sealed interface ExprDto : ValueDto
+
+@Serializable
+@SerialName("NewExpr")
+data class NewExprDto(
+    val classType: TypeDto,
+) : ExprDto {
+    override val type: TypeDto
+        get() = classType
+
+    override fun toString(): String {
+        return "new $type()"
+    }
+}
+
+@Serializable
+@SerialName("NewArrayExpr")
+data class NewArrayExprDto(
+    val elementType: TypeDto,
+    val size: ValueDto,
+) : ExprDto {
+    // TODO: support multi-dimensional arrays, when upstream implement it
+    override val type: TypeDto
+        get() = ArrayTypeDto(elementType, 1)
+
+    override fun toString(): String {
+        return "new $elementType[$size]"
+    }
+}
+
+@Serializable
+@SerialName("DeleteExpr")
+data class DeleteExprDto(
+    val arg: ValueDto,
+) : ExprDto {
+    override val type: TypeDto
+        get() = BooleanTypeDto
+
+    override fun toString(): String {
+        return "delete $arg"
+    }
+}
+
+@Serializable
+@SerialName("TypeOfExpr")
+data class TypeOfExprDto(
+    val arg: ValueDto,
+) : ExprDto {
+    override val type: TypeDto
+        get() = StringTypeDto
+
+    override fun toString(): String {
+        return "typeof $arg"
+    }
+}
+
+@Serializable
+@SerialName("InstanceOfExpr")
+data class InstanceOfExprDto(
+    val arg: ValueDto,
+    @SerialName("type") val checkType: TypeDto,
+) : ExprDto {
+    override val type: TypeDto
+        get() = BooleanTypeDto
+
+    override fun toString(): String {
+        return "$arg instanceof $checkType"
+    }
+}
+
+@Serializable
+@SerialName("LengthExpr")
+data class LengthExprDto(
+    val arg: ValueDto,
+) : ExprDto {
+    override val type: TypeDto
+        get() = NumberTypeDto
+
+    override fun toString(): String {
+        return "$arg.length"
+    }
+}
+
+@Serializable
+@SerialName("CastExpr")
+data class CastExprDto(
+    val arg: ValueDto,
+    override val type: TypeDto,
+) : ExprDto {
+    override fun toString(): String {
+        return "($type) $arg"
+    }
+}
+
+@Serializable
+@SerialName("PhiExpr")
+data class PhiExprDto(
+    val args: List<ValueDto>,
+    val blocks: List<Int>,
+) : ExprDto {
+    override val type: TypeDto
+        get() = args.first().type
+}
+
+@Serializable
+@SerialName("ArrayLiteralExpr")
+data class ArrayLiteralDto(
+    val elements: List<ValueDto>,
+    override val type: TypeDto,
+) : ExprDto {
+    override fun toString(): String {
+        return elements.joinToString(prefix = "[", postfix = "]")
+    }
+}
+
+@Serializable
+@SerialName("ObjectLiteralExpr")
+// TODO: keys and values
+data class ObjectLiteralDto(
+    // val keys: List<String>,
+    // val values: List<Value>,
```

Comment:
Why is it commented with TODO?

Reply:
Objects and arrays were (and probably are still) a mess in ArkIR. Commented code is the expected "good solutions". But it might not match ArkIR at all. We should re-check it.

Reply:
Turns out, objects are *much* more complex than just "a map with keys and values". There are 5 different "property element kinds": property assignment (`let obj = { a: 42 }`), shorthand property assignment (`let obj = { a }`), spread assignment (`let obj = { ...other }`), method declaration (`let obj = { method() {...} }`), accessor declaration (`let obj = { get accessor() {...} }`, also `set`). All of them are syntactic constructions and cannot be "simplified" into a list of 'key:value' pairs, since the final result can only be determined at the runtime (due to spread operator, for example).

I think we can design it that way on our side, but ultimately we have to wait for the support in ArkIR on the arkanalyzer side, and them adapt to it. Or force it by ourselves via PRs there.